### ISSUE TITLE
Domain disputes: special reserved list (dispute domains)

### ISSUE THREAD

**Issue by vohmar** [internetee/registry#269]
As of 1st of January 2017 the domains registrations that have been successfully disputed must be reserved for the contender for up to 3 years.

* new list (dispute domains) similar to reserved domains list (/admin/reserved_domains) - columns name, pw, created/updated at, valid to, actions. Filter fields - name, valid to from , valid to until, results per page
* domain form for adding domains to the list similar to reserved form (/admin/reserved_domains/new) with additions of mandatory 'valid to' and 'comment' (for url to dispute decision) fields
* registry must check that in case of epp domain update query with \<domain:chg>\<domain:registrant> element (ie registrant change) the domain is not in the dispute domains list - if it is in the list the \<eis:extdata>\<eis:reserved> element with correct password must be present in the extension block
* registry must check that in case of epp domain create query the domain name is not in the reserved nor dispute domains lists - if it is the \<eis:extdata>\<eis:reserved> element with correct password must be present in the extension block
* if a domain name that is in the reserved domains list is added to the dispute domains list reserved list password is replaced with the new password from dispute domains list (if a domain in the dispute list is added to the reserved list the password in the reserved list is set from the dispute list - admin needs to be notified if the optional password field had a value in it - "proposed password rejected, dispute password used instead") - password for any domain in both reserved and dispute lists must be the same (shared)
* if the password for a domain in dispute list is used for domain create or domain update the domain is deleted from the dispute list (in case of reserved domain - the reserved password is reset like in regular reserved domain registration)
* The domains in dispute list will get a dynamic disputed status in WHOIS identically to current reserved and blocked statuses. Described in sub-issues https://github.com/internetee/rest-whois/issues/29 and https://github.com/internetee/whois/issues/15
* If the contender does not register the domain within the period of 3 years since the dispute decision (valid to of dispute list record) the domain will be removed from dispute list.

---
To test:
1. WHOIS regeneration: on domain update
2. WHOIS regeneration: on contact update
3. WHOIS regeneration: manually via rake whois:regenerate
4. /registrant/whois
5. WHOIS regeneration: on domain delete
6. Dispute/Blocked Domain/Reserved Domain management

**Comment by artur-beljajev** [comment]
№ 3: What is the expected behaviour if validation fails?

**Comment by vohmar** [comment]
No3, 4: If reserved pw is missing for a domain in the dispute list an error must be returned
`2003: Required parameter missing; reserved pw element required for dispute domains`

**Comment by artur-beljajev** [comment]
> created/updated at
I don't see update action listed above, so probably there is nothing to track.

> If the contender does not register the domain within the period of 3 years since the dispute decision (valid to of dispute list record) the domain will be removed from dispute list.
automatically via CRON?

**Comment by artur-beljajev** [comment]
6 disputes per year

**Comment by artur-beljajev** [comment]
@vohmar What permissions should be set for dispute list?

**Comment by artur-beljajev** [comment]
How about leaving our "Results per page" feature? I can't imagine when it will be used

**Comment by artur-beljajev** [comment]
Does "password" field remain optional?

**Comment by artur-beljajev** [comment]
Can I implement domain selector as a drop-down, instead of a text field as it is in "reserved domains" section?

**Comment by vohmar** [comment]
1. the password should be editable (like in case of reserved list). Also the comment and valid to values should be editable - everything that is entered manually by admin.
2. cron is OK
3. results per can be left out from initial implementation (in case it is time consuming), but might be necessary for some other, bigger registry if one should be interested in using our registry software.
4. password field in optional in the sense that if left blank system will have to generate it (like in the case of reserved list)
5. yes, domain selector can be drop-down if it will not present performance issues with generating the values in the drop down

**Comment by artur-beljajev** [comment]
1. OK, but only the last update will be tracked. Does it seem helpful?

**Comment by artur-beljajev** [comment]
How about deletion? Is it also needed?

**Comment by vohmar** [comment]
1. better if we could track the whole history with dates and updator info
4. yes, password is always set and necessary
6. yes, comment field is free-text
7. yes, there can be only one effective decision per domain

**Comment by artur-beljajev** [comment]
№ 6: Are you sure single comment field will be enough? Perhaps it's worth implementing something more sophisticated, say like this: http://prntscr.com/e3kwu5?

The same could be applied to another entities as well (say, domain or contact)

**Comment by artur-beljajev** [comment]
№ 7: Do we need to validate that expiration date? For example to make impossible to use date in the past.

**Comment by artur-beljajev** [comment]
№ 8: Is pagination needed?

**Comment by artur-beljajev** [comment]
№ 9: What is the default order on dispute list page?

**Comment by vohmar** [comment]
6. yes, I do not see the need for extra details. The comment field is for storing reference to the dispute case, there is no need for detailed description in registry system
7. this is nice to have feature - helps admin from making one type of mistake
8. pagination is not necessity, but in the context of having unified views and reusable code it might also be a good idea to have it. Your decision - what ever looks good and is simple to implement
9. default order should be latest/newest records first

**Comment by artur-beljajev** [comment]
№ 10: Do you want to be able to delete entries?

**Comment by vohmar** [comment]
10. yes, ie to remove records with typos. There must be a "are you sure" sanity check.

**Comment by artur-beljajev** [comment]
#355 is needed for this ticket, so I need to merge it.

**Comment by artur-beljajev** [comment]
№ 11: Is this the correct response if the password is invalid?

    <result code="2202">
      <msg lang="en">Invalid authorization information; invalid reserved&gt;pw value</msg>
    </result>

**Comment by artur-beljajev** [comment]
№ 12: How about introducing separate dispute view? Otherwise all fields do not fit well in the list http://prntscr.com/e51taj

**Comment by artur-beljajev** [comment]
№ 13: Perhaps you would like to see some info about dispute in domain view as well? (/admin/domains/xxx)

**Comment by vohmar** [comment]
11. yes, that is correct error code and message
12. that is a good suggestion - on the list view lets have domain name, create/update date, expiration date, password, action buttons. On the details view there are separate created at and updated at fields, everything that is on the list view, comment field, current registration data (admin/domains/86919: statuses, general, registrant).

**Comment by vohmar** [comment]
13. Lets introduce new status for example serverDisputed (like serverForceDelete) that comes with set of sub-statuses in particular serverRenewProhibited, serverRegistrantChangeProhibited. If a domain is added to the dispute domain list the serverDisputed status together with its sub-statuses is applied automatically (need to check for existing conflicting statuses pendingUpdate and pendingRenew - UI should warn admin ui user asap during dispute domain list entry creation (when domain name is entered in the form) and adding record to the list should fail until conflicting statuses are resolved/removed). PS: list of sub-statuses needs additional analysis.

**Comment by vohmar** [comment]
replace create with update as on record creation update = create by default.

**Comment by artur-beljajev** [comment]
Note that the password is optional, even though there is an asterisk. It will be fixed by #367

http://prntscr.com/e5v5q6

**Comment by artur-beljajev** [comment]
Dispute list is under "Settings" > "Settings" menu.

**Comment by vohmar** [comment]
the decision on the statuses that must be present on domains on the dispute list is RegistrantChangeProhibited only

**Comment by artur-beljajev** [comment]
№ 6 (from description): There is some confirmation involved in case `request_confrimation_on_registrant_change_enabled` setting is true. A dispute should probably not be closed until that confirmation happens?

**Comment by vohmar** [comment]
@artur-beljajev yes that is correct as this involves pending registrant change that is in direct conflict with registrant change prohibition. This should already be solved with the existing statuses, so if setting registrant change prohibited status fails adding a record to dispute list must fail as well until conflicting statuses are resolved.

**Comment by artur-beljajev** [comment]
But when a dispute should be closed?

**Comment by vohmar** [comment]
domain is removed from the dispute list only after the transform request has been successfully processed ie data is changed in the registry - new domain record created in case of domain:create or registrant reference changed in case of domain:update. That means if domain:update should drop into pending state the record cannot be removed from the list before the action is confirmed and registrant object actually replaced.

**Comment by artur-beljajev** [comment]
Timo: the system does not support multiple reserved pw values on domain:create. So for this reason the reserved and dispute pw values should always be the same in case the domain is in both lists

**Comment by artur-beljajev** [comment]
WHOIS-related tickets require that disputed domain names are also present in WHOIS database. Is it OK?

**Comment by vohmar** [comment]
@artur-beljajev yes, the disputed domains should be in the whois database in separate table like in case of reserved and blocked domains.

**Comment by vohmar** [comment]
this branch cannot be automatically merged

https://github.com/internetee/registry/pull/373

**Comment by vohmar** [comment]
* please add validation to enable adding only domains in the zones that the registry serves. At the moment anything goes.
* add column name 'domain' on the first column of dispute list on admin
* remove the # and brackets from the domain name in the first column of disputes list
* need to add reserved pw field to the portal for registrars under edit menu (it is currently present only in create domain view)
* update epp doc (domain.md) domain update part to add reserved pw option

**Comment by vohmar** [comment]
* if domain is in both reserved and dispute lists and I change the pw in reserved list I get the success message "Reserved domain has been successfully updated" but the password is actually not changed. It should throw an error like "update failed, password from dispute list is in use"
* if user adds disputed domain to reserved list and sets password the message returned is "Reserved domain has been successfully created" while it should also explain that the password users chose for the reserved domain is disregarded and password from dispute list is used instead.

**Comment by artur-beljajev** [comment]
№ 2, 3: 1st column in dispute list looks like this for a reason: a link points to particular dispute view. "#1" denotes dispute's id, and domain name goes in brackets. I wouldn't add any name to this column since it only represents a link. I also consider the idea to just leave domain name there bad, since then it will look like a link to a domain, which is incorrect. Perhaps we should extract domain name (possibly making it a link pointing to a domain) into separate column and leaving dispute's id as a link.

**Comment by artur-beljajev** [comment]
№ 4: You mean here /registrar/domains/edit?domain_name=test.ee? But how it is connected to dispute? I tried to create a domain using the form at /domains/new and filling "Reserved pw" field, and it was saved in `domains.auth_info` column.

**Comment by vohmar** [comment]
no2, 3: the current form is noisy for admin as real dispute ids do not correspond to automatic numbers set by registry, so please remove the number and bracket thing. I see two options here - name the column 'dispute link' or name it domain and add action button 'details' for accessing and editing dispute data. 

no4: dispute commitee makes a decison about giving registration over from one registrant to the other. ie domain is usually registered when this decision is fulfilled. The reserved pw field in /registrar/domains/new is verified against the reserved domains password value. The same option must be available for registrant change in /registrar/domains/edit?domain_name=test.ee. Cannot comment on the names of variables code uses.

no6: it must be clear to the user why the option to change reserved domain pw is unavailable. So some kind of message is also necessary if error based approach is too complicated.

no7: obviously the current tip is not enough because I did not even notice it.

PS please also add column name for the last column called 'Actions' similarly to the portal for registrars

**Comment by artur-beljajev** [comment]
№ 2, 3: First column now serves as a link to a dispute, at the same time displaying domain name.

**Comment by vohmar** [comment]
merge conflicts: https://github.com/internetee/registry/pull/410

**Comment by artur-beljajev** [comment]
Merged

**Comment by vohmar** [comment]
1. When adding domain from unmanaged zone the error message should state "invalid domain name" or "domain is not managed by this registry" or "domain name is in unsupported zone"
2. Domain edit view in the portal for registrars (/registrar/domains/edit?domain_name=...) is still missing an option to add reserved pw
3. Action heading on top of the last column is still nice to have to keep unified stile with other view (ie reserved list) and the portal for registrars that do have such heading for the actions column.

**Comment by vohmar** [comment]
4. domains do not get removed from the dispute list when the exp date has arrived - I have domains in the list with exp dates back in february
5. please rename the exp. date column/field to smth more descriptive ie "dispute resolution valid to"

**Comment by artur-beljajev** [comment]
1. Fixed
2. Added
3. Removed from everywhere
4. CRON on staging has been updated manually
5. Renamed to "valid to"

**Comment by artur-beljajev** [comment]
Please test the following whois regeneration cases:
1. Domain update
2. Contact update
3. Manually via rake whois:regenerate
4. Dispute/Blocked Domain/Reserved Domain add/edit/remove

**Comment by vohmar** [comment]
1. /admin/disputes - rename valid to range filter - valid to: from ... to ...
2. So sorry, but please remove the registrantChangeProhibited status from dispute domains - the password protection is enough
3. font color of column headings of the dispute list is incorrect - needs to be blue like in other tables.

**Comment by artur-beljajev** [comment]
Please notice new test case above (№6)

**Comment by artur-beljajev** [comment]
1. Renamed
2. No problem, removed
3. Blue color denotes link, There are no links there, what's why the color is different
4. Fixed

**Comment by artur-beljajev** [comment]
There is a bug: if you change dispute's domain name, whois record of old name remains and new record for new name is generated. Should I handle both names (remove old one, insert new) or just prohibit domain name change in dispute form?

**Comment by vohmar** [comment]
block the name update - should be simpler to implement.

**Comment by vohmar** [comment]
dispute list is missing history - so once a disputed domain is registered to the new owner record is deleted from the db. If record is deleted from the disputes table is must be moved or the copy of it must be kept in history table ie log_disputes with the record of the "delete" date.

**Comment by artur-beljajev** [comment]
№ 1: Is some UI needed to be able to browse deleted records?

**Comment by artur-beljajev** [comment]
Blocked by #536

**Comment by vohmar** [comment]
#1 yes admin must have access to the "deleted" dispute records

**Comment by artur-beljajev** [comment]
domains.reserved

**Comment by vohmar** [comment]
1. disputed status is missing - not present in registry/epp not whois
2. restwhois ei toimi - uurin martiniga mis seal viga võib olla
3. please use reserved element for disputed password - adding disputed element means development for registrars and that might be a problem
4. domain update / edit view needs an additional field to enter the reserved password - ideally this field should be dynamic, available only if the domain in question is disputed (consider effort - this is nice to have)
5. if verified yes element is missing from domain update the update request is put in bending but domain is already reamoved from the disputed list. If request is not confirmed wihting 48 hours the pending requet is canceled, previous owner remains as registrant but domain is free of disputed restrictions. Prefered solution here would be to autoconfirm reigstrant change id correct password is provided.
6. if admin deletes the disputed record it is not moved to expired / closed disputes list
7. there should not be a delete button for expired/closed disputes - as we do not have archive table for disputes (ie log_disputes) we need to save history. Instead we should show the date when the record was removed from active dipsutes list and if possible together with reason (registered/expired/deleted)
8. in the add to disputed view under start at cell there is a comment "Must be at least today / in future" that is not true. User can enter any date in the field. Actually the date should be today or in the past becuase domain should not be inserted to disputed list unless the dispute commitee's decision is in force. But this gives an alternative use case of not adding any manual statuses to prohibit unwanted updates to the registration. Will discuss this with registry admin.

**Comment by karlerikounapuu** [comment]
@vohmar 
> disputed status is missing - not present in registry/epp not whois

It's implemented just as reserved domain, eg. it's manipulated directly in WhoisRecord. Should it be viewable / editable as a status in registry as well? If it should, then registry admin must keep track of expiry dates herself because it's automated only in case Dispute model is created.

As per current implementation it should be present in whois/rest-whois. True, *-disputed-status branch (in rwhois/whois) should be applied at time.

> restwhois ei toimi - uurin martiniga mis seal viga võib olla

This PR additionally required changes in rest-whois end. It should be tested in conjuction with rwhois/whois *-disputed-status branch. Otherwise it may trigger error about serializing.

**Comment by vohmar** [comment]
Comments:
1. We might indeed escape the need to add disputed status on registry level, whois could be enough just like in case of reserved and blocked domains. But there is a slight difference here - while in case of reserved domains the reserved status is there only in case the domain is not registered with disputed domains the status should possibly be always there even if the domain is registered to note that registrant change is restricted. Whois and rest-whois records' updating needs some mapping to understand why the statuses are not there in staging
2. is fixed now - rest-whois is working on staging

Additional issues:
9. removing a domain from disputed list while it is not registered should put the domain in auction list (if the listing expires or admin deletes it)

**Comment by vohmar** [comment]
State after new complete test run:
4. the password field was still missing from domain edit view (for domains in disputed list)
9. disputed and deleted domain was not put into auction after expiry of disputed status
10. Its possible to add domains into disputed list that are not registered
11. the DisputeStatusUpdateJob does not log or return any results to stdout - it should at minimum declare how many domains were archived and preferably with domain names that were removed from active disputed list.

additional comments and ideas:
* please rename the menu option for domain disputes to disputed domains to mach the style of other lists ie reserved domains and blocked domains
*expired/closed disputes list in admin - remove password column, replace expires at column with expired/closed at column, add updator column to mark who/what removed the record from diputes (expire job, admin, registrant change, registration)

**Comment by vohmar** [comment]
Latest rest results:
4. the password field did not show when error message was returned ie for not adding password for registrant change
11. job logs results but returns incorrect information. An example:
```
I, [2020-05-21T13:10:47.768349 #11673]  INFO -- : DisputeStatusUpdateJob - Found 2 closable disputes
DEPRECATION WARNING: PaperTrail.whodunnit is deprecated, use PaperTrail.request.whodunnit (called from add_updator at /home/registry/deploy/releases/20200520150349/app/models/concerns/versions.rb:26)
I, [2020-05-21T13:10:48.211055 #11673]  INFO -- : DisputeStatusUpdateJob - closed dispute  for 'timoajalugu10.ee'
I, [2020-05-21T13:10:48.276470 #11673]  INFO -- : DisputeStatusUpdateJob - Failed toclose dispute for 'timoajalugu9.ee'
I, [2020-05-21T13:10:48.277901 #11673]  INFO -- : DisputeStatusUpdateJob - Found 0 activatable disputes
I, [2020-05-21T13:10:48.279222 #11673]  INFO -- : DisputeStatusUpdateJob - All done. Closed 1 and activated 0 disputes.
I, [2020-05-21T13:10:48.279482 #11673]  INFO -- : DisputeStatusUpdateJob - Failed to close disputes with Ids:[20]
```
in reality both domains were successfully removed form dispute list. The supposedly failed timoajalugu9 was successfully sent to auction. 

12. any domain update request (unless there is registrant element with current id present in the chg section) returns error requiring reservedpw. In reality only registrant change should be prohibited without correct password
